### ISSUE TITLE
Hardcoded ValidUntil causing issues, SAML2 standard references it as optional

### ISSUE THREAD

**Issue by Glowsome** [onelogin/php-saml#502]
Case : 

- Microfocus AccessManager IDP v4.5.3
- OneLogin-PHP Client on SP.

Issue:

- The IDP does not have a function to dynamically reload Metadata.
- As validUntil is hardcoded in \lib\Saml2\Metadata.php this is causing issues.
- Current (unwanted) Workaround is manually removing the entry from the generated Metadata.
- According [https://docs.oasis-open.org/security/saml/v2.0/saml-metadata-2.0-os.pdf](https://docs.oasis-open.org/security/saml/v2.0/saml-metadata-2.0-os.pdf) Line 316 the parameter is **optional**

Solution wanted:

- As stated above the definition of validUntill should be an Optional parameter.
- This will solve manual interaction on the Metadata before importing it into the mentioned IDP.

**Comment by pitbulk** [comment]
Is not hardcoded, you can provide it as a parameter:
- https://github.com/onelogin/php-saml/blob/master/lib/Saml2/Settings.php#L897
- https://github.com/onelogin/php-saml/blob/d074a814c45b96fd69b0401493385a17e006c533/lib/Saml2/Metadata.php#L27

**Comment by Glowsome** [comment]
> Is not hardcoded, you can provide it as a parameter:
> 
>     * https://github.com/onelogin/php-saml/blob/master/lib/Saml2/Settings.php#L897
> 
>     * https://github.com/onelogin/php-saml/blob/d074a814c45b96fd69b0401493385a17e006c533/lib/Saml2/Metadata.php#L27

You are correct, However if it **NOT** defined it gets set with a default value, so whatever i do .. its not vanishing.

Please see [https://github.com/onelogin/php-saml/blob/master/lib/Saml2/Metadata.php#L30](https://github.com/onelogin/php-saml/blob/master/lib/Saml2/Metadata.php#L30)

As it is then in the Metadata by default - see [https://github.com/onelogin/php-saml/blob/master/lib/Saml2/Metadata.php#L159](https://github.com/onelogin/php-saml/blob/master/lib/Saml2/Metadata.php#L159)
There is no way to get rid - other then hacking into the generate metadata manually.

**Comment by Glowsome** [comment]
Additional argument of having it **not** present if it is undefined in settings is the fact that if an IDP is not capable of dynamically reloading Metadata (as it is in my case) the default set expiration (as written in the code) sets it to 2 days.

So in this : 
- To avoid setting/defining it to some ridiculous setting/timespan (like 2099) it should be not present in the Metadata itself.
- A possible solution (if the tag cannot be ommited, and the parameter is not explicitly set) is to Query the certificate's expiry date from the IDP, and then dynamically set the ValidUntill parameter to the same as the expiry of the certificate, therefore aligning it.

**Comment by pitbulk** [comment]
If I change the behavior of the current method that generates the Metadata to not include such values if not provided, I can break other environments that expects it

**Comment by Glowsome** [comment]
> If I change the behavior of the current method that generates the Metadata to not include such values if not provided, I can break other environments that expects it

I am not really sure as to interpret this reply so i will state it harshly ... ?
Either you are suggesting :
- "changing behaviour as to what is 'optional' as the standard dictates is unconveint to you"
or
- "i wave off your arguments and just continue on this 'hardcoded-way'"

Either way i have forked the repo, and am trying to correct this (for me unwanted) way of implemenatation to get rid on this IMHO ugly hack to manually strip the metadata of the optional tags without compromising the IDP's who can coap with it.

- Glowsome

**Comment by pitbulk** [comment]
From the SAML spec:  https://docs.oasis-open.org/security/saml/v2.0/saml-metadata-2.0-os.pdf  Page 9

> When used as the root element of a metadata instance, this element (<EntitiesDescriptor> or <EntityDescriptor>) MUST 
> contain either a validUntil or cacheDuration attribute. It is RECOMMENDED that only the root element of a metadata instance
> contain either attribute.

So you need to set an expiration on the SAML Metadata, that why when not provided, I set default expiration times.

> The IDP does not have a function to dynamically reload Metadata.
This is your specific case of use. Is the IdP reading the SP Metadata directly from stored XMLs rather than process the XML when it is received and storing the parameters? 

>"changing behaviour as to what is 'optional' as the standard dictates is unconveint to you"

As I shared, a Metadata with no expiration does not follow the standard, that why I'm not changing the current behavior.

But in addition, if the toolkit has an specific behavior that the developers expect, is not that easy to change it, it need to be done in a major release when it can impact current developers integrations.  Currently php-saml has 9.3M of downloads at [packagist](https://packagist.org/packages/onelogin/php-saml), so I need to take care of them, and keep the integration stable.

You had the option to fork the project (as you did), but there are other alternatives like [simplesamlphp](https://github.com/simplesamlphp/simplesamlphp) or [lightSAML](https://github.com/lightSAML/lightSAML)

**Comment by Glowsome** [comment]
> This is your specific case of use. Is the IdP reading the SP Metadata directly from stored XMLs rather than process the XML when it is received and storing the parameters?

The XML/Metadata is retrieved when creating the connection ( either retrieval via url or file upload )
There is no mechanism in place to dynamically reload/re-grab metadata other then a manual action on the IDP itself.

Also the refrenced page, and text are relevant when using Entitie**s**Descriptor, not the actual EntityDescriptor where its all about.

Yes i do concur that when Entitie**s**Descriptor is used then the 'MUST' should be honored as written in the standard.

- Glowsome

**Comment by pitbulk** [comment]
I disagree, the section is talking about |EntitiesDescriptor| or |EntityDescriptor| [One or More]

**Comment by Glowsome** [comment]
> I disagree, the section is talking about |EntitiesDescriptor| or |EntityDescriptor| [One or More]
> 
> And the [saml2int](https://kantarainitiative.github.io/SAMLprofiles/saml2int.html#_metadata_and_trust_management) profile directly set validUntil as mandatory.
> 
> > [SDP-MD03]
> > Metadata without a validUntil attribute on its root element MUST be rejected.

Again you are quoting me on information that refers to having <EntitiesDescriptor> as Root element. ( secttion 2.3.1 - line 310 thru 358.
In which it is stated that when <EntitiesDescriptor> is the root element then ValidUntill is a MUST.

However ... 
If you were to read the next section ( 2.3.2 Element <EntityDescriptor>  Lines 360 thru 442)
The 'MUST' is not found regarding the ValidUntill, but only mentioned as options.

**Comment by Glowsome** [comment]
@pitbulk if i were to get this done, would you be willing to concider a PR if i were to submit one ?

I mean my aim is to honor both the current implementation and be able to also enable the ones who use my type/brand of IDP to be able to use the already provided code - so in essense i am trying to broaden the implementation.

I am not aiming to move away from current code just by forking, but enhance ( and theerefore gain support) thru the general code to also support my case i would very much welcome my effort.
In this i would (really) like to get a 1:1 convo going as to where we can get to a mutual agreement in implementing this without affecting current users, but does enhance it to the thing (my IDP) is in need of.

- Glowsome

**Comment by pitbulk** [comment]
Maybe the less intrusive way is to add a "$noExpiration" (by default to false) parameter to the Metadata [builder](https://github.com/onelogin/php-saml/blob/master/lib/Saml2/Metadata.php#L27) and then create in the method a 
```
$expirationTimeSection= "";
if (!$noExpiration) {
    $expirationTimeSection = """validUntil="{$validUntilTime}"
                     cacheDuration="PT{$cacheDuration}S"
                      """;
}
```
and then in the template
```
<md:EntityDescriptor xmlns:md="urn:oasis:names:tc:SAML:2.0:metadata"
              {$expirationTimeSection}entityID="{$spEntityId}">
```
